Add string operator enlist-input[]

saqimtiaz opened the pull request:
While [title lists](https://tiddlywiki.com/#Title%20List) are very common in TiddlyWiki, we are missing a filter operator that allows it's input to be parsed as a title list.

For example consider `[{!!parent}get[children]]` where `children` is a field holding a title list. 
Currently this requires two steps to operate on the list of children, first getting `children` and then `enlist<children>`
```
<$vars children={{{ [{!!parent}get[children]] }}}/>
<$list filter="[enlist<children>]">
<!-- do something -->
</$list>
</$vars>
```

With this operator the above would be:

`[{!!parent}get[children]aslist[]]`

I've considered extending `enlist[]` to operate on its input when no operand is present, but that might be confusing for users.

I'm open to better suggestions for a name:
- `parselist[]`
- `tolist[]`
- `titlelist[]`
- ?

See [this table](https://tiddlywiki.com/#Filter%20Operators) for names of existing string operators.

BurningTreeC commented:
@saqimtiaz I believe this helps me a lot

BurningTreeC commented:
@saqimtiaz - I've encountered a problem, where a filter is passed as a variable to another filter. The variable can contain both a filter or a text reference to a filter. I'd like to have an operator that leaves the filter output as it is if the variable is a filter and parses the text reference to get the filter behind it if it's a text reference. Is this filter doing just that?

saqimtiaz commented:
@BurningTreeC no this is a string manipulation filter to turn a string like:
`"[[Title A]] TitleB [[Title C]]" `
into a list which can be manipulated with listops or iterated over:
```
"Title A"
"TitleB"
"Title C"
```

We can currently do this with `enlist[]` but if the string is the result of previous filter steps we can't chain them, leading to the need for an extra step with a set/vars widget.

pmario commented:
I think `aslist[]` is a good name

Jermolene commented:
Thanks @saqimtiaz I agree that we should not extend the "enlist" operator to cover this use case. Nonetheless, it might be helpful for users to understand that this is the same operator made available in two different packages. So, decorating the name "enlist" in some way might be a good approach, especially if there are other operators that we wish worked on input titles instead of their operand.

"enlist-input[]" could be a candidate?

Jermolene commented:
Hi @saqimtiaz sorry, another thought: I think it might be worth adding some tests for this operator to clarify the behaviour of edge cases (eg what happens when there are multiple input titles).

saqimtiaz commented:
@Jermolene I was trying to maintain the naming pattern of string operators (trim, split etc) but from a user perspective your suggestion makes more sense.

I have updated the PR to use `enlist-input`

As for multiple inputs, this is a string operator so the behaviour is the same as for other string operators like `split`, the output is the union of the output for each input title.

I am not familiar with the test framework we use but can look into it.

saqimtiaz commented:
@Jermolene tests added and filter renamed to `enlist-input`.

I'll prepare a separate PR with docs once this is merged.

Jermolene commented:
Thanks @saqimtiaz

rmunn commented:
#5151 is probably important to consider: I think `enlist-input` should have "dedupe" and "raw" suffixes like `enlist` does, to let the user choose whether or not it should dedupe its input. Right now the behavior appears to be deduplication, with no way to select "raw" mode.

rmunn commented:
PR #5152 adds "dedupe" and "raw" suffixes to `enlist-input`, similar to `enlist`.
